### ISSUE TITLE
Upstream data-exhaustion hang

### ISSUE THREAD

**Issue by sdi1100041** [sandyresearch/parcae#12]
**The symptom.** In a multi-rank run, whichever rank holds the least data exhausts its
shards first. When it does, it leaves the training loop without raising; the remaining
ranks then block in the next collective until the NCCL watchdog aborts them ~30 min
later. Rank 0 has already printed its completion message, so the job finishes with
**exit 0** and `sacct` reports `COMPLETED`. We lost four multi-hour runs to this before
realising the runs had not actually completed.

### Why it is silent

Three things compose, all on current `main`:

1. **No wraparound.** `scripts/train.py:1017` hardcodes `repetitions=None` for the
   `pqds`/`pqds-pure` path (the `hfds` path a few lines above does honour
   `curr_config.repetitions`), so a rank's stream is finite.

2. **Exhaustion is a normal loop exit, not an error.**
   `recpre/huggingface_dataset.py:208-212`, `ParquetStreamPure.__iter__`:

   ```python
   def __iter__(self):
       if not self.stateful:
           self._state_init()

       while self._state["file_idx"] < len(self.parquet_files):
   ```

   When the file list is used up the condition simply becomes false and the generator
   returns. Nothing raises, and nothing is communicated to the other ranks.

3. **The new packing wrapper propagates it.** `PackedIterableDataset.__iter__` in
   `recpre/data_loading_utils.py` (added by #11) does the same thing:

   ```python
       while True:
           collator._refill(doc_iter)
           if not collator.doc_buffer:
               return
   ```

The rank that returns falls out of the training `for` loop while its peers advance to the
next step and wait for it in DDP's gradient all-reduce. A checkpoint save cannot rescue
the situation either, because `ParquetStreamPure.state_dict()` itself does a
`torch.distributed.all_gather` (`recpre/huggingface_dataset.py:264`) and so needs every
rank present.

### It is reproducible and predictable

Our corpus is 576 shards / 33.3B tokens; a 29.6B-token run at 16 ranks needs 1.850B
tokens per rank. Attributing tokens to shards from the corpus manifest and replaying the
sharding code, the unluckiest rank at `seed: 1337` gets 1.7247B — a 6.8% deficit,
predicting exhaustion at step **52,634** of 56,457. The run died at step **52,563** on
four separate attempts (0.13% from the prediction), each time reporting exit 0. At
`seed: 42` the same rank gets 1.8962B (+2.5%) and the run completes. So roughly half of
arbitrary seeds lose the run in its final stretch, and the failure looks like success.

### Why #11 does not fix it

#11 makes the per-rank file sets a true partition, which is correct and welcome — but a
partition equalises the *number* of shards per rank, not the number of *tokens*. Shard
sizes in our corpus have sd = 33% of the mean (1.34M to 65.65M tokens), so per-rank token
totals still differ by ~±25%. Replaying both code paths over 300 base seeds against the
same 29.6B budget:

| | seeds where every rank has enough | coverage |
|---|---:|---:|
| before #11 | 159 / 300 | ~63% |
| after #11 | 140 / 300 | 100% |

The difference between 159 and 140 is not meaningful — the point is that it is still
about half either way. Concretely, `seed: 42`, the one value we had empirically proven to
complete a run, flips from +2.5% to **−3.9%** under the new assignment, because the
assignment changed. So the hazard is unchanged by #11, and anyone who had found a working
seed will need a different one.

### Suggested fix

Any one of these would remove the silent-failure mode; the first is the cheapest:

- **Raise instead of returning** when a rank exhausts its shards, so the job dies with a
  clear message instead of hanging. Even better, have ranks agree: all-reduce a
  "still have data" flag each step and stop everyone together at the first exhaustion.
- **Honour `repetitions` for `pqds`** (or default it to wrap) so the stream is infinite
  and short-changed ranks simply repeat.
- **Assign shards by token count rather than by count**, using row-group counts or file
  sizes, so per-rank totals are balanced to within a shard.
- Cheap belt-and-braces: at startup, compare each rank's available tokens against
  `max_steps * micro_batch_size * block_size * grad_accum` and fail fast if any rank is
  short.

### Environment

`main` @ `69284c1` (with #11 merged). Lines cited from that commit. Reproduced originally
on `6e45195` with local modifications, but none on the code paths above — the three
mechanisms are unmodified upstream code in both.

**Comment by Hprairie** [comment]
Yeah, this is a fair point, and I see what you are saying. I don't have the full bandwidth to fix this right now, but I can get around to it eventually.

I have always been operating in a regime where I had much more pretraining data available (i.e., using Climbmix 400B and Fineweb-Edu 350B) and thus haven't been affected. 

The fix to this seems non-trivial to me; raising an error on data repetition doesn't seem like a good idea, and exiting seems better. Honour repetitions for pqds seems like a valid thing to allow people to do. Assigning shards by token count is too complicated and creates extra intro overhead, as we need to preprocess. The cheap belt-and-braces also doesn't work, as it involves preprocessing.

I agree that we should honor repetitions for pqds. So I can put it on my list of things to do. However, if you want to make sure that training completes, the appropriate pipeline would be to preprocess the entire dataset and pack the data. Then do something like this:

```yaml
all_block_size_tensors: true
pack_sequences: false

data_config:
  train_data:
    - type: pqds-pure
      data_dir: path/to/train
      prefix: ""
      data_signature:
        keys: ["input_ids"]

```

We can then round-robin shards and then add logic to shard rows in case there is unequal sharding.
